Implement sending commands to solver in incremental SMT2 backend

thomasspriggs opened the pull request:
This PR includes implementation of sending commands to solver in incremental SMT2 backend, based on the exprts recieved. This is sufficient to get a sat/unsat response from the solver in sufficiently simple cases. However parsing the result from the solver is not included in this PR and will be in a follow up PR.
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] N/A - Non claimed. ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Make decision procedure use piped solver

thomasspriggs committed:
Update existing test to add correct z3 arguments

So that the z3 sub process is actually started and expecting smt2
commands from the pipe in this test.

thomasspriggs committed:
Send initialisation commands to solver

thomasspriggs committed:
Add debug logging of commands sent to SMT2 solver

So that we can see the text sent to the solver when debugging and so
that the text can be checked in regression tests. The commands are shown
as debug output messages rather than written to file because a file
based approach is not well suited to incremental solving.

thomasspriggs committed:
Add test of the setup commands sent to smt2 solver

thomasspriggs committed:
Implement symbol_exprt to smt_termt conversion

thomasspriggs committed:
Add namespace access to `smt2_incremental_decision_proceduret`

So that instances of `symbol_exprt` can be handled correctly.

thomasspriggs committed:
Add conversion of exprs to SMT2 and sending to solver process

thomasspriggs committed:
Implement `dec_solve`

thomasspriggs committed:
Fix guard identifier processing

The identifier for guards includes the `\` character. Therefore this
commit is needed for allowing smt identifiers to be constructed
including this character and converted to string.

thomasspriggs committed:
Add test of an example with control flow

In order to check that guards are converted ok.

TGWDB commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -12,6 +13,9 @@ smt2_incremental_decision_proceduret::smt2_incremental_decision_proceduret(
     number_of_solver_calls{0},
     solver_process{split_string(solver_command, ' ', false, true)}
 {
+  send_to_solver(smt_set_option_commandt{smt_option_produce_modelst{true}});
+  send_to_solver(smt_set_logic_commandt{
+    smt_logic_quantifier_free_uninterpreted_functions_bit_vectorst{}});
```
Do we want to send these at solver construction time, or do we want to send/update these per query? I don't have a good answer now, but there was comment on another issue/PR that the logic could vary for different calls/inputs. Also do we want to be able to change these later (and is this allowed/easy to do)?

thomasspriggs replied:
According to section 4.1 of the SMTLIB standard version 2.6, the `set-logic` command can only be issued whilst the solver is in "Start mode" and the only way to return to this mode is to issue the `reset` command. The actions of the `reset` command include undefining all functions and assertions we have sent to the solver. Therefore if we were reseting and setting different logics for different rounds of solving, then we would no longer be carrying out incremental solving.

TLDR; We should be sending these commands at construction time.

NlightNFotis reviewed (approved):
LGTM

peterschrammel commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -8,10 +8,12 @@
+  : solver_command(std::move(_solver_command)),
     number_of_solver_calls{0},
-    solver_process{split_string(solver_command, ' ', false, true)}
+    solver_process{split_string(solver_command, ' ', false, true)},
```
The split_string looks brittle.

thomasspriggs replied:
The arguments are provided to `cbmc` in space separated form. Therefore they currently need to be split in order separate them. Any suggestions for making this less brittle?

peterschrammel commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -2,11 +2,93 @@
+
+/// \brief Defines any functions which \p expr depends on, which have not yet
+///   been defined, along with their dependencies in turn.
+void smt2_incremental_decision_proceduret::define_dependent_functions(
```
Would be great to unit-test this somehow.

thomasspriggs replied:
I have added unit tests of `smt2_incremental_decision_proceduret` these tests indirectly test the functionality of this `define_dependent_functions` function, via the `handle` and `set_to` functions of the public interface.

thomasspriggs committed:
Disable running incremental smt2 tests on Windows

These tests currently fail when run on Windows due to issues with
`piped_processt`. This commit can be reverted once the debugging is
complete.

thomasspriggs committed:
Refactor to separate solver communication from decision process

So that `smt2_incremental_decision_proceduret` can be unit tested in
isolation using a mock smt solver process.

thomasspriggs committed:
Add smt2_incremental_decision_proceduret unit tests

thomasspriggs committed:
Fix handling of symbols with no value

Because there was previously an assumption that all symbols in the
symbol table had values and it is more straight forward to make the this
code deal with that than to add an invariant.

thomasspriggs committed:
Refactor symbol lookup

The previous implementation contained 2x "not" operations. By appling De
Morgan's laws these operations can be dispensed with and the code is
thus made more readable.

NlightNFotis commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -67,21 +67,21 @@ void smt2_incremental_decision_proceduret::define_dependent_functions(
       const irep_idt &identifier = symbol_expr->get_identifier();
       const symbolt *symbol = nullptr;
-      if(!ns.lookup(identifier, symbol) && !symbol->value.is_nil())
+      if(ns.lookup(identifier, symbol) || symbol->value.is_nil())
```
Are we sure this change is valid? It doesn't look like that to me.

[De Morgan's law in this particular instance would be a bi-implication](https://en.wikipedia.org/wiki/De_Morgan%27s_laws):

```
¬(A ∨ B) ⇔ ¬A ∧ ¬B
```

Assuming in your case `A : ns.lookup()` and `B : symbol->value.is_nil()` then I believe this should have been translated to `(!(ns.lookup() || symbol->value.is_nil()))`.

NlightNFotis replied:
Yeah, I just verified this is with a quick run of the truth table:

`A : ns.lookup()`
`B : symbol->value.is_nil()`
`Original : !ns.lookup() && !symbol->value.is_nil()`
`Refactored : ns.lookup() || symbol->value.is_nil()`

| A  | B | Original | Refactored |
|---|---|--------|-------------|
| F  | F  |      T      |         F           |
| F  | T  |       F     |         T           |
| T  | F  |       F     |         T           |
| T  | T  |       F     |        T            |

thomasspriggs replied:
Yes. The outcome of the logic you present is the opposite way round. That final not was handled by flipping the contents of the code blocks rather than adding the `not` operation. Therefore the outcome is still the same, which is why the tests still pass.

NlightNFotis reviewed (changes requested):
The rest of the stuff looks good, but the refactoring of the condition expression changed its semantics.

thomasspriggs committed:
Improve error printing in smt2 decision procedure unit test

In order to improve the chances of being able to see what is wrong
when/if the tests fail.

thomasspriggs committed:
Make the order of SMT commands for sub-expressions deterministic

This makes the behaviour of the conversion predictable. This has the
advantage of being easier to unit test and reducing the possiblity of
hard to re-produce bugs.

chrisr-diffblue commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -2,16 +2,133 @@
+        return expr_identifier.first;
+      });
+  std::stack<exprt> to_be_defined;
+  const auto dependencies_needed = [&](const exprt &expr) {
```
Any particular reason why this is a lambda, rather than a helper function, such as a hypothetical `push_dependencies_needed(exprt, std::stack<exprt>)` which might be more descriptive? Either way, I'd be tempted to change the name to make its behaviour more explicit... At the moment, it superficially looks like a call to `dependencies_needed(e)` would be state-free - but actually its changing the state of the method via the stack of exprts... I'd make that behaviour explicit in the name for ease of code reading...

thomasspriggs replied:
Because it was the quickest way for me to de-duplicate this code in the 2 places it is needed. A separate function would also have needed an additional parameter. I agree that the name I assigned to this lambda isn't necessarily the best. I think doing some pair programming together could be helpful in trying to come up with a more readable implementation.

chrisr-diffblue commented on src/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -50,7 +50,7 @@ bool smt_bool_literal_termt::value() const
 static bool is_valid_smt_identifier(irep_idt identifier)
 {
-  static const std::regex valid{R"(^[^\|\\]*$)"};
+  static const std::regex valid{R"(^[^\|]*$)"};
```
Minor quibble - but for the more hieroglyphic regex's like this its nice to include a comment showing an/some examples of what the regex is attempting to match.

thomasspriggs replied:
I agree this would be better with a comment. I didn't add one when I first wrote this, as it is relatively short and I am relatively comfortable with regex. I'll try and remember to comment this in a follow-up PR.

chrisr-diffblue reviewed (approved):
Nothing blocking - though I've left a couple of small comments that might be worth considering.
